File: leetcode91/leetcode91_my_version.cpp
class Solution {
public:
    int numDecodings(string s) {
        int n = s.length();
        vector<int> ways(n, 0);
        // invalid
        if (n == 0 || s[0] == '0')
            return 0;
        
        
        if (n >= 1 && s[0] != '0') {
            ways[0] = 1;
        }
        if (n >= 2) {
            if (s[0] < '2' && s[1] != '0')
                ways[1] = 2;
            else if (s[0] <= '2' && s[1] == '0')
                ways[1] = 1;
            else if (s[0] == '2' && s[1] <= '6')
                ways[1] = 2;
            else if (s[0] == '2' && s[1] >= '7')
                ways[1] = 1;
            else if (s[0] >= '3' && s[1] != '0')
                ways[1] = 1;
            // invalid
            else if (s[0] >= '3' && s[1] == '0')
                return 0;
        }
        
        for (int i = 2; i < n; ++i) {
            if (s[i] != '0')
                ways[i] += ways[i - 1];
            
            if (s[i - 1] > '0' 
                && (s[i - 1] < '2' || (s[i - 1] == '2' && s[i] <= '6')))
                ways[i] += ways[i - 2];
            
            // can't decode
            if (ways[i] == 0)
                return 0;
        }
        
        return ways[n - 1];
    }
};

// Runtime: 4 ms, faster than 65.13% of C++ online submissions for Decode Ways.
// Memory Usage: 8.6 MB, less than 47.06% of C++ online submissions for Decode Ways.
// time complexity: O(n)
// space complexity: O(n)

File: leetcode91/leetcode91_my_version.java
class Solution {
    public int numDecodings(String s) {
        int n = s.length();
        int[] ways = new int[n];
        
        // base case
        // xxxxxxx | 5
        if (s.charAt(n - 1) != '0')
        {
            ways[n - 1] = 1;
        }
        // xxxxxx |2 | 5
        // xxxxxx | 25
        if (n > 1 && s.charAt(n - 2) != '0')
        {
            if (Integer.valueOf(s.substring(n - 2, n)) <= 26)
            {
                ways[n - 2]++;
            }
            
            ways[n - 2] += ways[n - 1];
        }
        
        for (int i = n - 3; i >= 0; i--)
        {
            // general case
            // xxxxxxx |3| 25 => |3|2|5| or |3|25|
            // xxxxxxx |32 | 5
            int counts = 0;
            if (s.charAt(i) != '0')
            {
                // xxxxxxxx | 3 25 => |2|5| or |25|
                counts += ways[i + 1];
                // xxxxxxxxx |32 |5. 32 should <= 26
                if (Integer.valueOf(s.substring(i, i + 2)) <= 26)
                {
                    counts += ways[i + 2];
                }
            }
            
            ways[i] = counts;
        }
        
        return ways[0];
    }
}

// Runtime: 1 ms, faster than 93.45% of Java online submissions for Decode Ways.
// Memory Usage: 37.7 MB, less than 38.18% of Java online submissions for Decode
// Ways.
// time complexity: O(n)
// space complexity: O(n)